platform/src/main/java/org/hillview/main/BatchLogAnalysis.java

```diff
@@ -97,48 +111,54 @@ private static HeatmapData heatmapErrTime(FileSetDescription desc, int numOfTime
-                        sb.append(heatmapData.timeLabels.get(x));
-                        sb.append(',');
-                        sb.append(heatmapData.errorCodeLabels.get(y).replaceAll(",",""));
-                        sb.append(',');
-                        sb.append(heatmapData.matrix[x][y]);
-                        sb.append('\n');
+        if (!heatmapData.errorCodeLabels.contains(null)) {
+            System.out.println(heatmapData.errorCodeLabels);
+            try (PrintWriter writer = new PrintWriter(new File(filepath))) {
+                StringBuilder sb = new StringBuilder();
+                sb.append("Timestamp");
+                sb.append(',');
+                sb.append("errorCode");
+                sb.append(',');
+                sb.append("Count");
+                sb.append('\n');
+                for (int x = 0; x < heatmapData.matrix.length; x++) {
+                    for (int y = 0; y < heatmapData.matrix[0].length; y++) {
+                        if (heatmapData.matrix[x][y] >= 0) {        // keep zeros or not
+                            sb.append(heatmapData.timeLabels.get(x));
+                            sb.append(',');
+                            sb.append(heatmapData.errorCodeLabels.get(y).replaceAll(",", ""));
+                            sb.append(',');
+                            sb.append(heatmapData.matrix[x][y]);
+                            sb.append('\n');
+                        }
                     }
                 }
+                writer.write(sb.toString());
+            } catch (FileNotFoundException e) {
+                System.out.println(e.getMessage());
+                System.exit(1);
             }
-            writer.write(sb.toString());
-        } catch (FileNotFoundException e) {
-            System.out.println(e.getMessage());
-            System.exit(1);
         }
     }
 
     /**
      * This method generates and saves heatmapData to CSV files given input syslogs.
-     * @param desc description of files of RFC5424 log format
      * @param logDir source directory of syslogs for all bugs
      * @param figDir destination directory to save the CSV files for all bugs
      * @param numOfTimestampBuckets number of timestamp (x-axis) intervals for each heatmap
      */
-    public static void getBugHeatmaps(FileSetDescription desc, String logDir, String figDir, int numOfTimestampBuckets) {
+    public static void getBugHeatmaps(String logDir, String figDir, int numOfTimestampBuckets) {
         File path = new File(logDir);
         String[] bugIDs = path.list(new FilenameFilter() {    // each subDir in logDir corresponds to one bug
             @Override
             public boolean accept(File current, String name) {
                 return new File(current, name).isDirectory();
             }
         });
+        // create empty directory for nsx_manager and nsx_edge respectively
+        new File(figDir + "/nsx_manager_syslog/").mkdirs();
```

Comment:
Do we know that this is actually useful?

Reply:
Separating nsx_manager and nsx_edge should be useful. The number of distinct errorCodes in each one should be smaller.